e2e/gpu/gpu_test.go

```diff
@@ -62,10 +69,121 @@ func TestGPU(t *testing.T) {
 		pods, err := ct.Kubeclient.PodsFromDeployment(ctx, ct.Namespace, gpuDeploymentName)
 		require.NoError(err)
 		require.Len(pods, 1, "pod not found: %s/%s", ct.Namespace, gpuDeploymentName)
+		pod = &pods[0]
+	})
+
+	t.Run("check podvm->container mounts by libnvidia-container", func(t *testing.T) {
+		require := require.New(t)
+		assert := assert.New(t)
+		require.NotNil(pod, "require 'wait for GPU deployment' to succeed, no pod found")
+		ctx, cancel := context.WithTimeout(t.Context(), ct.FactorPlatformTimeout(3*time.Minute))
+		defer cancel()
+
+		expectBins := []string{
+			// Binaries taken from output of libnvidia-container v1.17.8.
+			"nvidia-smi",
+			"nvidia-debugdump",
+			"nvidia-cuda-mps-control",
+			"nvidia-cuda-mps-server",
+		}
+		for _, bin := range expectBins {
+			for _, cmd := range []string{
+				fmt.Sprintf("[[ $(command -v %s) == /usr/bin/%s ]]", bin, bin),
+				fmt.Sprintf("test -x /usr/bin/%s", bin),
+			} {
+				argv := []string{"/usr/bin/env", "bash", "-c", cmd}
+				stdout, stderr, err := ct.Kubeclient.Exec(ctx, ct.Namespace, pod.Name, argv)
+				assert.NoError(err, "running %q:\nstdout:\n%s\nstderr:\n%s", cmd, stdout, stderr)
+			}
+		}
+
+		const libPath = "/usr/lib/x86_64-linux-gnu"
+		expectLibs := map[string]struct {
+			abiLink   bool // Wether to expect a link with ABI version, like .so.1
+			unverLink bool // Wether to expect a link without any version, like .so
+		}{
+			// Libraries taken from output of libnvidia-container v1.17.8.
+			"libnvidia-ml.so":              {abiLink: true},
+			"libnvidia-cfg.so":             {abiLink: true},
+			"libcuda.so":                   {abiLink: true, unverLink: true},
+			"libcudadebugger.so":           {abiLink: true},
+			"libnvidia-opencl.so":          {abiLink: true},
+			"libnvidia-gpucomp.so":         {abiLink: false},
+			"libnvidia-ptxjitcompiler.so":  {abiLink: true},
+			"libnvidia-allocator.so":       {abiLink: true},
+			"libnvidia-pkcs11-openssl3.so": {abiLink: false},
+			"libnvidia-nvvm.so":            {abiLink: true},
+		}
+		for lib, libChecks := range expectLibs {
+			pathThisLib := path.Join(libPath, lib)
+
+			// Run `ls` to check what libraries with that name exist.
+			getLibsCmd := fmt.Sprintf("ls %s*", pathThisLib)
+			stdout, stderr, err := ct.Kubeclient.Exec(ctx, ct.Namespace, pod.Name, []string{"/usr/bin/env", "bash", "-c", getLibsCmd})
```

Comment:
Did you consider running one `find` on the libPath, outside the loop, and then processing the results locally? I'm a bit worried that the large number of `Kubeclient.Exec` calls might increase flakiness.

Reply:
Thought about it, but it looked like it worked quite well (and fast) during development, and it helps a bit with debugability and readability of the code.
I'm fine with switching the ls in loop to a single find, but we still call exec 1-3 times for each lib, so we will still do many execs.

We could also do all this in a bash script that we copy over for the test, but I thought the feedback we get in ci is better this way.

Reply:
Ok, then let's go ahead and see.